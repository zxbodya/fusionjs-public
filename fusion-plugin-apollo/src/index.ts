/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */
import type {DocumentNode} from 'graphql';
import ApolloRenderEnhancer from './plugin';

export * from './tokens';
export * from './apollo-client/index';

export {ApolloRenderEnhancer};

export function gql(path: string): DocumentNode {
  throw new Error('fusion-plugin-apollo/gql should be replaced at build time');
}
