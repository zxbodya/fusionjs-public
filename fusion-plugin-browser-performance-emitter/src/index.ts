/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import Server from './server.js';
import Browser from './browser.js';

declare var __BROWSER__: Boolean;
const BrowserPerformanceEmitter = __BROWSER__ ? Browser : Server;

export default BrowserPerformanceEmitter;
