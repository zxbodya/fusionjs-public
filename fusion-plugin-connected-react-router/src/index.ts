/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import {createPlugin} from 'fusion-core';
import {applyMiddleware, compose} from 'redux';

import {
  routerMiddleware as createRouterMiddleware,
  connectRouter,
} from 'connected-react-router';
import {RouterToken} from 'fusion-plugin-react-router';

import type {ConnectedRouterPluginType} from './types';

const plugin: ConnectedRouterPluginType = createPlugin({
  deps: {
    router: RouterToken,
  },
  provides: ({router}) => {
    const enhancer = (createStore) => {
      return function _createStore(reducer, initialState) {
        const store = createStore(reducer, initialState);
        // $FlowFixMe - We enhance the store to add ctx onto it, which doesn't exist in the redux libdef
        const {history} = router.from(store.ctx);
        const routerReducer = connectRouter(history);
        const combinedReducer = (
          {router: routerState, ...restState} = {},
          action
        ) => {
          /**
           * `reducer` is likely the result of `combineReducers`, which
           * warns if you pass it more state than it expects
           */
          return {
            ...reducer(restState, action),
            router: routerReducer(routerState, action),
          };
        };
        const newStore = createStore(
          combinedReducer,
          initialState,
          compose(applyMiddleware(createRouterMiddleware(history)))
        );
        return newStore;
      };
    };
    return enhancer;
  },
});

export {ConnectedRouterEnhancerToken} from './tokens';
export default plugin;
