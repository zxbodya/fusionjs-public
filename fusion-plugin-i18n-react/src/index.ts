/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import I18n from './plugin';

export {
  I18nToken,
  I18nLoaderToken,
  I18nTranslateFnsToken,
  HydrationStateToken,
  createI18nLoader,
} from 'fusion-plugin-i18n';
export type {
  I18nDepsType,
  I18nServiceType,
  TranslationsObjectType,
  TranslateFuncType,
} from 'fusion-plugin-i18n';

export {withTranslations} from './with-translations';
export {useTranslations} from './use-translations';
export {Translate} from './translate';
export {I18nContext} from './plugin';

export default I18n;
