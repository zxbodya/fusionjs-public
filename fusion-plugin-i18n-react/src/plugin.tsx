/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import * as React from 'react';
import PropTypes from 'prop-types';

import i18n, {I18nToken} from 'fusion-plugin-i18n';
import type {I18nDepsType, I18nServiceType} from 'fusion-plugin-i18n';
import {FusionContext, ProviderPlugin, useService} from 'fusion-react';

type ExtractReturnType = <V, TArg>((arg: TArg) => V) => V;
export type I18nType = $Call<
  ExtractReturnType,
  $PropertyType<I18nServiceType, 'from'>
>;
export const I18nContext = React.createContext<I18nType>({});

/**
 * The i18n service is loaded with the fusion ctx and provided to the
 * application. This is not explicitly necessary from an API perspective, but it
 * is critical that we register a callback to splitComponentLoaders for
 * dynamically loading translations.
 */
function BundleSplitConsumer(props, {splitComponentLoaders}) {
  const ctx = React.useContext(FusionContext);
  const service: I18nServiceType = useService(I18nToken);
  const i18n = service.from(ctx);
  // Keep track of i18n instance in case of hot-reloading
  const [callbacks, setCallbacks] = React.useState({
    i18n,
    // `i18nKeys` comes from fusion-react/async/split
    load: (_, {i18nKeys}) => i18n.load(i18nKeys),
  });
  // `splitComponentLoaders` comes from fusion-react/async/prepare-provider
  // eslint-disable-next-line react-hooks/exhaustive-deps
  React.useMemo(() => splitComponentLoaders.push(callbacks.load), []);
  React.useMemo(() => {
    if (i18n !== callbacks.i18n) {
      const index = splitComponentLoaders.find((fn) => fn === callbacks.load);
      const load = (_, {i18nKeys}) => i18n.load(i18nKeys);
      splitComponentLoaders.splice(index, 1, load);
      setCallbacks({i18n, load});
    }
  }, [i18n]); // eslint-disable-line react-hooks/exhaustive-deps
  return (
    <I18nContext.Provider value={callbacks.i18n}>
      {React.Children.only(props.children)}
    </I18nContext.Provider>
  );
}

BundleSplitConsumer.contextTypes = {
  splitComponentLoaders: PropTypes.array,
};

export default ProviderPlugin.create<I18nDepsType, I18nServiceType>(
  'i18n',
  i18n,
  BundleSplitConsumer
);
