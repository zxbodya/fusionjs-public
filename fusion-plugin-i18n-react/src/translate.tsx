/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import React from 'react';

import {I18nContext} from './plugin.js';

type TranslatePropsType = {
  id: string,
  data?: Object,
};

function Translate(props: TranslatePropsType) {
  const i18n = React.useContext(I18nContext);
  const content =
    (i18n && i18n.translate && i18n.translate(props.id, props.data)) ||
    props.id;
  return React.Fragment ? <>{content}</> : <span>{content}</span>;
}

export {Translate};
