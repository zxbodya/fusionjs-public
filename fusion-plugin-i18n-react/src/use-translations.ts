/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import {useContext, useMemo} from 'react';
import {I18nContext} from './plugin';

export function useTranslations() {
  const i18n = useContext(I18nContext);
  if (!i18n || !i18n.translate) {
    throw new Error(
      'Could not find i18n React context. Perhaps the context provider is missing (i.e. fusion-plugin-i18n-react was not registered) or the provider and useTranslations are referencing different instances of the fusion-plugin-i18n-react React context (i.e. multiple instances of fusion-plugin-i81n-react in the dependency tree).'
    );
  }
  return useMemo(() => i18n.translate.bind(i18n), [i18n]);
}
