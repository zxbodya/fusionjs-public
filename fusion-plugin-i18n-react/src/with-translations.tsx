/** Copyright (c) 2018 Uber Technologies, Inc.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import React from 'react';
import type {Context} from 'fusion-core';
import {I18nContext} from './plugin.js';

type TranslateType = (
  key: string,
  interpolations?: {[string]: string | number}
) => string;

/*
The `withTranslations` HOC takes an array of translation keys as an argument,
but does not use it at runtime.
However, these keys are captured by `babel-plugin-i18n` at compile-time by
Fusion's compiler and the compiler uses generate a map of all translations
in the app.
The translation map is then exposed by `fusion-plugin-i18n/chunk-translation-map.js`
*/

export const withTranslations = (
  translationKeys: string[] /*translationKeys*/
) => {
  return <T: {}>(
    Component: React$ComponentType<T>
  ): Class<
    React$Component<
      $Diff<T, {|translate?: TranslateType, localeCode?: string|}>
    >
  > => {
    class WithTranslations extends React.Component<T> {
      translate: TranslateType;
      localeCode: string;

      constructor(props: T, context: Context) {
        super(props, context);
        const i18n = context;
        this.localeCode = i18n && i18n.localeCode ? i18n.localeCode : 'en_US';
        this.translate =
          i18n && i18n.translate
            ? (key: string, interpolations?: {[string]: string | number}) =>
                i18n.translate(key, interpolations)
            : (key: string) => key;
      }

      render() {
        return (
          <Component
            {...this.props}
            translate={this.translate}
            localeCode={this.localeCode}
          />
        );
      }
    }

    const displayName = Component.displayName || Component.name || 'Anonymous';
    WithTranslations.displayName = `withTranslations(${displayName})`;
    // $FlowFixMe
    WithTranslations.contextType = I18nContext;

    return WithTranslations;
  };
};
