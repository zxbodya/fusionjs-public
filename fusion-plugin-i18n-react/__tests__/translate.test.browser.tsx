// @flow

import React from 'react';
import {render} from '@testing-library/react';

import {Translate} from '../src/index';
import {I18nContext} from '../src/plugin.js';

test('Translate', () => {
  const Foo = () => {
    return (
      <div>
        <Translate id={'foo'} />
      </div>
    );
  };

  const mockI18n = {
    async load() {},
    localeCode: 'fr_CA',
    translate() {
      return 'foo bar baz';
    },
  };

  const {asFragment} = render(
    <I18nContext.Provider value={mockI18n}>
      <Foo />
    </I18nContext.Provider>
  );

  expect(asFragment()).toMatchSnapshot();
});
