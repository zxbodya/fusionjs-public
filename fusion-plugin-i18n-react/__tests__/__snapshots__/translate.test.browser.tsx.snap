// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Translate 1`] = `
<DocumentFragment>
  <div>
    foo bar baz
  </div>
</DocumentFragment>
`;
